### ISSUE TITLE
Negative pseudo-R2 on the training set

### ISSUE THREAD

**Issue by cxrodgers** [glm-tools/pyglmnet#232]
First off, thanks all for the amazing work that has gone into this package. I'm really excited to use this to analyze my data, it fills a big hole in the existing resources that are out there.

I'm using this to fit a Poisson GLM to explain spike counts using a design matrix with 750K rows and ~100 feature columns. A lot of the features are dummy binary indicators for categorical variables. For demonstration purposes, alpha=0 and reg_lambda=0. I get a negative pseudo-R2 (about -.04) when I score the model on the training set (the same data that was used to fit the model). Probably I'm just doing something stupid but I don't understand how this is even possible. It's my understanding that, at a minimum, the model should be able to predict the mean firing rate and achieve a pseudo-R2 of zero.

Here's my code:
```
glm = pyglmnet.GLM(distr='poisson', alpha=0, reg_lambda=[0], score_metric='pseudo_R2', verbose=True)
glm.fit(exog.values, endog.values)
```

This prints:

> Looping through the regularization path
> Lambda: 0.0000
> 	Converged. Loss function: 0.04
> 	dL/L: -0.000996

The score is negative on the training set:

> In [13]: glm.score(exog.values, endog.values)
> Out[13]: array([-0.04439965])

What can I try to debug this?

I would be happy to provide the data if it would be helpful but the matrices are quite large so let me know if that is necessary. `exog` is a pandas DataFrame with zero mean and unit variance. `endog` is a pandas Series of spike counts (4318 spikes total, mean .006 spikes per bin).

**Comment by jasmainak** [comment]
hi @cxrodgers can you try lowering the convergence tolerance to see if it helps? I discovered some issues with the stopping criteria recently

**Comment by cxrodgers** [comment]
Hi @jasmainak I think you are right that this is the issue, I think it reported convergence too early. 

Lowering `tol` substantially improves the fit. However it also starts to run very slowly. Changing to solver='cdfast' makes it run much faster, and it also seems to make the result less sensitive to changes in `tol`. Is there any reason not to use 'cdfast'?

It seems my dataset is very sensitive to changes in `tol`. Do you think you could give me any quick guidance on how to appropriately choose `tol`? I wonder if I could implement a different stopping criterion based on how quickly pseudo-R2 is converging, rather than dL/L, would that make sense?

Thanks!!

**Comment by jasmainak** [comment]
I think the convergence criterion should be based on the loss, not on the scoring metric. However, I am indeed quite curious if we could use another stopping criteria. One option is to use the norm of the gradient, so you break the loop when the algorithm stops making progress. An advantage of this is that you do not need to compute the loss in every loop which in itself is expensive. Would you like to try implementing this and see if it works better on your dataset (and on the examples in pyglmnet)?

Regarding the default as `'cdfast'`, I don't have a clear answer. Perhaps @pavanramkumar can through some light on this.

**Comment by pavanramkumar** [comment]
@cxrodgers I'm happy that you're using this for your data. 

I've now made a WIP pull request to update convergence criteria from using loss to gradient norm. It also has a fix for early / premature convergence. 

As @jasmainak said above, it's faster than loss-based metrics because we don't need to explicitly compute the loss.

See #233  

I'd be curious to know if this helps you achieve good fits without having to tune `tol` very much.

**Comment by cxrodgers** [comment]
Sounds great! I think we can close this issue because lowering the tolerance to 1e-6 (and raising the max number of iterations to 1e6) fixed this problem for me.

I'm interested in the new stopping criteria. I will give it a shot when I upgrade from the pip version to the github version.

Thanks!
